bulk_indexer.go

```diff
@@ -322,35 +336,37 @@ func (b *bulkIndexer) Flush(ctx context.Context) (BulkIndexerResponseStat, error
 				b.retryCounts[b.itemsAdded] = count
 
 				if b.gzipw != nil {
-					gr, err := gzip.NewReader(bytes.NewReader(b.copyBuf))
-					if err != nil {
-						return resp, fmt.Errorf("failed to decompress request payload: %w", err)
+					// First loop, read from the gzip reader
+					if len(buf) == 0 {
+						n, _ := gr.Read(buf[:cap(buf)])
+						buf = buf[:n]
 					}
-					defer gr.Close()
 
-					// keep track of the previous newlines
-					// the buffer is being read lazily
-					seen := 0
+					newlines := bytes.Count(buf, []byte{'\n'})
 
-					n, _ := gr.Read(buf[:cap(buf)])
-					buf = buf[:n]
-					// loop until we've seen the nth newline
-					for newlines := bytes.Count(buf, []byte{'\n'}); seen+newlines < startln; newlines = bytes.Count(buf, []byte{'\n'}) {
+					// loop until we've seen the start newline
+					for seen+newlines < startln {
 						seen += newlines
 						n, _ := gr.Read(buf[:cap(buf)])
 						buf = buf[:n]
+						newlines = bytes.Count(buf, []byte{'\n'})
 					}
 
 					startIdx := indexnth(buf, startln-seen, '\n') + 1
 					endIdx := indexnth(buf, endln-seen, '\n') + 1
 
 					// If the end newline is not in the buffer read more data
-					for endIdx == 0 {
-						// Add more capacity (let append pick how much).
-						buf = append(buf, 0)[:len(buf)]
+					if endIdx == 0 {
+						// loop until we've seen the end newline
+						for seen+newlines < endln {
+							// Add more capacity (let append pick how much).
+							buf = append(buf, 0)[:len(buf)]
```

Comment:
This seems a bit inefficient given the initial buf size of 1024 bytes. A simple test like `TestAppenderRetryDocument/gzip` needs to grow this buffer 20 times

Without knowing more, I'd suggest using much more than 1024 bytes as initial capacity; maybe 16k? - we have so few bulkIndexers that it shouldn't make a material difference